iroh-dns-server/src/store/signed_packets.rs

```diff
@@ -42,73 +167,53 @@ impl SignedPacketStore {
     }
 
     pub fn open(db: Database) -> Result<Self> {
+        // create tables
         let write_tx = db.begin_write()?;
-        {
-            let _table = write_tx.open_table(SIGNED_PACKETS_TABLE)?;
-        }
+        let _ = Tables::new(&write_tx)?;
         write_tx.commit()?;
-        Ok(Self { db: Arc::new(db) })
+        let (send, recv) = mpsc::channel(1024);
+        let cancel = CancellationToken::new();
+        let cancel2 = cancel.clone();
+        let actor = Actor {
+            db,
+            recv,
+            cancel: cancel2,
+            max_batch_size: MAX_BATCH_SIZE,
+            max_batch_time: MAX_BATCH_TIME,
+        };
+        // start an io thread and donate it to the tokio runtime so we can do blocking IO
+        // inside the thread despite being in a tokio runtime
+        let handle = tokio::runtime::Handle::try_current()?;
+        let thread = std::thread::Builder::new()
+            .name("packet-store-actor".into())
+            .spawn(move || {
+                handle.block_on(actor.run());
+            })?;
```

Comment:
I think it will. The docs for Handle::block_on say:

"Runs a future to completion on this Handle’s associated Runtime.

This runs the given future on the *current thread*, blocking until it is complete, and yielding its resolved result. Any tasks or timers which the future spawns internally will be executed on the runtime."

Also, as far as I can see the `F` type parameter in block_on does not require Send, so tokio isn't even allowed to move the actual execution to one of the main tokio thread pool threads.

The reason I like this more than creating a separate current_thread or local runtime is that if you use spawn_blocking or spawn inside the fn, this is perfectly fine and will use the normal blocking or non-blocking thread pool of the main runtime.

Reply:
Here is a rust playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d35c78b6fa8dddd7d314067354d2de40

You can move an `Rc<...>` into the block_on, so it must be polled on the same thread.